webapp/src/components/App/Studies/StudyTree.tsx

```diff
@@ -21,18 +21,127 @@ import TreeItemEnhanced from "../../common/TreeItemEnhanced";
+ * they shoudln't wait for a long running scan to complete before they're able to
+ * a folder in the hierarchy.
+ *
+ * @param studiesTree
+ * @param folder
+ * @returns
+ */
+function mergeStudyTreeAndFolder(
+  studiesTree: StudyTreeNode,
+  folder: NonStudyFolder,
+): StudyTreeNode {
+  if (folder.parentPath == studiesTree.path) {
+    for (const child of studiesTree.children) {
+      if (child.name == folder.name) {
+        // parent path is the same, folder name is the same
+        // we don't override the existing folder
+        // so we return here and don't update this node
+        return studiesTree;
+      }
+    }
+    // parent path is the same, but no folder with the same name at this level
+    return {
+      ...studiesTree,
+      children: [
+        ...studiesTree.children,
+        {
+          path: `${folder.parentPath}/${folder.name}`,
+          name: folder.name,
+          children: [],
+        },
+      ],
+    };
+  } else if (folder.parentPath.startsWith(studiesTree.path)) {
+    // recursively walk though the tree
+    // recursively merge in each child
+    return {
+      ...studiesTree,
+      children: studiesTree.children.map((child) =>
+        mergeStudyTreeAndFolder(child, folder),
+      ),
+    };
+  } else {
+    // folder isn't part of this hierarchy
+    return studiesTree;
+  }
+}
+
+/**
+ * Merge several folders in the study tree.
+ *
+ * @param studiesTree
+ * @param folders
+ * @returns
+ */
+export function mergeStudyTreeAndFolders(
+  studiesTree: StudyTreeNode,
+  folders: NonStudyFolder[],
+): StudyTreeNode {
+  let mergedStudyTree = studiesTree;
+  for (const folder of folders) {
+    mergedStudyTree = mergeStudyTreeAndFolder(mergedStudyTree, folder);
+  }
+  return mergedStudyTree;
+}
+
+async function fetchAndMergeSubfolders(
+  path: string,
+  studiesTree: StudyTreeNode,
+): Promise<StudyTreeNode> {
+  if (path === "root") {
+    // Under root there're workspaces not subfolders
+    return studiesTree;
+  }
+  const pathParts = path.split("/");
+  if (pathParts.length < 2) {
+    return studiesTree;
+  }
+  // path parts should be ["root", workspace, "foler1", ...]
+  const workspace = pathParts[1];
+  const subPath = pathParts.slice(2).join("/");
+  const subFolders = await api.getFolders(workspace, subPath);
+  const nextStudiesTree = mergeStudyTreeAndFolders(studiesTree, subFolders);
+  return nextStudiesTree;
+}
 
 function StudyTree() {
+  const emptyNode: StudyTreeNode = { name: "", path: "", children: [] };
+  const [studiesTree, setStudiesTree] = useState(emptyNode);
+  const [initializedFlag, setInitializedFlag] = useState(false);
   const folder = useAppSelector((state) => getStudyFilters(state).folder, R.T);
-  const studiesTree = useAppSelector(getStudiesTree);
+  const initialStudiesTree = useAppSelector(getStudiesTree);
   const dispatch = useAppDispatch();
 
+  useEffect(() => {
+    if (!initializedFlag) {
```

Comment:
Why you don't update when the `studiesTree` change after the first initialization?

Otherwise you can do this and remove the `useEffect`:

```typescript
const initialStudiesTree = useAppSelector(getStudiesTree);
const [studiesTree, setStudiesTree] = useState(initialStudiesTree);
```

Reply:
@smailio the bug of the subfolders not appearing on the initial render comes from that, here's what I suggest to complete with Samir's comment:

```typescript
  const initialStudiesTree = useAppSelector(getStudiesTree);
  const [studiesTree, setStudiesTree] = useState(initialStudiesTree);
  const folder = useAppSelector((state) => getStudyFilters(state).folder, R.T);
  const dispatch = useAppDispatch();

  // Initialize folders once we have the tree
  useEffect(() => {
    if (!folder || !initialStudiesTree.children.length) {
      return;
    }

    // extract folder paths excluding root
    const getFolderPaths = (path: string): string[] =>
      path.split("/").reduce<string[]>((acc, _, i, parts) => {
        const currentPath = parts.slice(0, i + 1).join("/");
        return currentPath !== "root" ? [...acc, currentPath] : acc;
      }, []);

    // sequential folder fetching with error handling
    const updateFolderTree = async () => {
      try {
        const finalTree = await getFolderPaths(folder).reduce(
          async (treePromise, path) => {
            const tree = await treePromise;
            return fetchAndMergeSubfolders(path, tree);
          },
          Promise.resolve(initialStudiesTree),
        );

        setStudiesTree(finalTree);
      } catch (error) {
        // here you can use a snackback error component if needed
        console.error("Failed to initialize folders:", error);
      }
    };

    updateFolderTree();
  }, [folder, initialStudiesTree]);
  ```